### PR TITLE
Runtime stitching APIs and sanity tests, ttnn runtime submit refactor

### LINKED ISSUE #671: Runtime stitching / runtime weights
Proposal for Front-End (FE) Interaction with the tt-mlir Runtime for On-Device Tensors

In this document, I will refer to the tt-mlir runtime as the “runtime” and the forge and pjrt runtimes as “third-party runtimes” (TPRT).

Key Use Cases:
- Multi-batch instance: Avoid rewriting weights/constants for every iteration.
- Runtime stitching: Output of one model feeds as input to another (including past cache), without feeding IOs back through the host.
- Training loops: Updatable weights on the device during training iterations.

The following, I believe, addresses all these use cases:
- TPRT should push parameters once, and they should remain live on the device.
- TPRT pushes activations for each iteration.
- TPRT can create two input tensors for double-buffered graph execution if needed.
- TPRT is responsible for deallocating all tensors.
- The runtime leaves outputs on the device until explicitly copied to the host by TPRT.
- Outputs can be stored in either L1 or DRAM.
- The compiler may not always know the layout of an on-device tensor (e.g., the FE may not recompile the same graph when executed twice, and activations could be in DRAM on the first execution and L1 cache on the second).
- If there is a layout mismatch, the runtime converts the tensor to the required layout.
- Tensors in L1 that are not needed by the current program should be moved to DRAM.

To accomplish this, I propose the following API:

-------------------

Tensor toDevice(Tensor, Device, Layout)
- Copies a tensor to the device with the specified layout.
- Returns a handle to the on-device tensor.

-------------------

Tensor toDevice(Tensor, Device)
- Copies a tensor to the device, interleaved into DRAM.
- Returns a handle to the on-device tensor.

-------------------

Tensor toHost(Tensor)
- Waits for tensor operations to complete and copies a tensor to the host.
- The tensor remains allocated in device memory.

-------------------

void wait(Tensor)
- Barriers on tensor operations being finished

-------------------

Layout getLayout(Binary, ProgramIndex, InputIndex)
 -Returns layout of input at index as defined in binary

-------------------

std::vector<Tensor> submit(Device, Binary, programIndex, inputTensors)  
- Execute binary on device
- Asserts all input tensors are on device
- Non-blocking, immediately returns, caller can barrier if desired
- Calls toLayout on input tensors that are not in correct layout
- Moves any tensors from L1 to DRAM that are not used by this program
- Returns list of on-device output tensors

-------------------

Tensor toLayout(Device, Tensor, Layout)
- Returns tensor of required layout

-------------------

void Deallocate(Tensor, Device)
- Deallocates tensor on device

### PR DESCRIPTION
**Pull request by jnie-TT**
closes #671 #103 #366 
Runtime support for APIs that enable stitching separate runtime submit calls together.

**APIs**
std::vector<Tensor> submit(Device, Binary, programIndex, inputTensors)  
- Execute binary.
- Implicitly executes asynchronously in ttnn unless blocking APIs are executed within the Binary graph such as ttnn::from_device.
- Returns an a vector of tensors.
- inputTensors can be on host or on device. We can save perf cycles if for example an input tensor is already on device. This way to_device ops can return immediately.
- Tensors returned may be on host or on device, depending on the flatbuffer execution graph.
- Any further blocking api calls such as toHost will automatically wait for all required tensors to be populated.
- Automatically converts input tensors to the desired layout.
-------------------

Layout getLayout(Binary, ProgramIndex, InputIndex)
 - Returns layout of input at index as defined in binary.
 
-------------------

Tensor toLayout(Tensor, Device, Layout)
- Returns tensor of required layout.
- TODO: Some paths that require host fallback for device tensor layout conversions are unchecked currently. We can add checks for that once we hit a use-case. Leaving unchecked for now.

-------------------

Tensor createTensor(Device device, Layout layout,
                    std::vector<std::uint32_t> const &shape,
                    std::vector<std::uint32_t> const &stride,
                    std::uint32_t itemsize)
                    
- Creates an empty tensor on host/device (location derived from layout) with the desired layout.

-------------------

void deallocateTensor(Tensor, bool force)
- Deallocates tensor. Set force to true to force deallocation.

-------------------

void memcpy(Tensor dst, Tensor src)
- Copies src to dst. Assumes bost tensors are on host and asserts dst size == src size.
- Works for host to host, host to device, device to host.
- TODO: add support for multichip tensors

-------------------

Tensor toHost(Tensor, bool untilize)
- Waits for tensor operations to complete and copies a tensor to the host.
- The original tensor remains allocated in device memory.
- Set untilize to true to untilize the tensor.

-------------------

**Testing**
- Updated ttrt and test_subtract to use the new submit API.
- Added a runtime python test folder that contains unit tests for runtime APIs. Added testing for the runtime stitching APIs in `runtime/test/python/ttnn/test_runtime_api.py`.
   - Currently this requires ttrt whls to be compiled and installed.
- Added a utility library for runtime testing. This currently contains helper functions that generate specific layouts, as we currently cannot get these layouts from the flatbuffers with `getLayout` until compiler support is added.
- Added a runtime stitching test that stitches an multiple eltwise programs together. Currently the mlir for this test is purely in the ttnn dialect and I manually removed to_device ops to avoid hitting asserts that the tensor is already on device, since the compiler currently hardcodes input layouts to host and generates ops based on this assumption.
   - On this note, we want to add an mlir directory specifically for runtime testing. Currently I added a folder under `test/ttmlir/Runtime` that contains the hacked ttnn dialect mlir for the runtime stitching test. Adding it under Silicon didn't work because ttrt will run all flatbuffers generated there, and the runtime stitching mlir needs a different flow. I'm open to suggestions for better locations of adding this, but it would be great to have a dedicated location to add mlir files that the compiler may not support yet and are used specifically for runtime testing.

**Backwards Compatibility**
- Kept the old submit API so that FEs have a grace period, but it would desirable that FEs migrate soon. 

**Misc**
- Fixed bug in isfinite tests where the input-output data formats are different. Currently the output of this test is incorrect on main.
- Marked simple_broadcast.mlir as unsupported as currently the generated ttnn graph is invalid: #1314

### CODE REVIEW

**Review by vmilosevic — APPROVED** [review]

File: runtime/include/tt/runtime/runtime.h
Comment (nsmithtt) [inline runtime/include/tt/runtime/runtime.h:94]: We should always pass the runtime API types by value because they wrap shared_ptr's and the intention is that they are reference counted.  Not saying that we ever should or would, but potentially we could make this API host async, i.e. feed it to some thread or worker queue in which case we'd want to keep a refcount to the API objects.

File: runtime/include/tt/runtime/runtime.h
Comment (nsmithtt) [inline runtime/include/tt/runtime/runtime.h:93]: I'm actually not sure we should return a `std::map`, I think a `std::vector` will suffice.  The `global_id` is only global to the flatbuffer binary internally, so there's nothing that the FE could actually do to make sense of this id.

A vector is better because then the FE can use the output index as an ID to correlate tensors between programs.

File: runtime/lib/runtime.cpp
Comment (nsmithtt) [inline runtime/lib/runtime.cpp:421]: This will break the FEs, can we have a grace period where both APIs are temp supported?

**Comment by nsmithtt** [comment]
@pilkicTT / @rpavlovicTT, would really like your input on these runtime API changes.

File: runtime/lib/runtime.cpp
Comment (jnie-TT) [reply runtime/lib/runtime.cpp:421]: Sure I'll keep the old API but log warning that it will be deprecated soon.

**Comment by rpavlovicTT** [comment]
> @pilkicTT / @rpavlovicTT, would really like your input on these runtime API changes.

At the moment it looks good overall.

Few general questions -- 
1. What about error handling? If something crashes/asserts (for example we run OOM) in runtime how does TPRT handles it?
2. Does it make sense to have `AllocateTensor` API as well? From the TPRT perspective, is there a case when you want to allocate memory but not to copy data immediately?  
3. Do we want to have callback support? As in, invoke a provided callback when an event is complete?
4. Should checkpointing be part of this API discussion or should it be separate?

**Comment by pilkicTT** [comment]
> @pilkicTT / @rpavlovicTT, would really like your input on these runtime API changes.

@nsmithtt Sorry, I haven't had a chance yet to take a look. Hopefully, later today or tomorrow (in the worst case)...

@jnie-TT I've noticed there are multiple unrelated changes in this PR, this should be avoided, especially with larger changes. For at least two reasons:
1. Making review harder
2. Making harder to revert the change, i.e. if runtime stitching creates a bug, you will need to manually revert (in order to keep other changes) or vice versa.

I am not blocking on this, just think it's good to avoid these situations in the future. Thanks!

**Comment by rpavlovicTT** [comment]
Thanks.  
I missed the callback support, kudos to @tapspatel . 

When you say
>  If we want to query info from the tensor those APIs will be blocking implicitly. 
Are you saying we'll be able to poll if tensor is ready/not? 

>
> 2. We can add this API if we need it, just wondering if it's sufficient to create the tensor once you have the data, or do we > have a use-case where pre-allocating the tensor is required?

I asked this question but don't have a concrete answer, it's more of a question for TPRT folks. cc @mrakitaTT @AleksKnezevic

File: runtime/lib/ttnn/runtime.cpp
Comment (pilkicTT) [inline runtime/lib/ttnn/runtime.cpp:367]: why this change?

File: runtime/lib/ttnn/runtime.cpp
Comment (pilkicTT) [inline runtime/lib/ttnn/runtime.cpp:459]: The other variant of `toDevice()` has an assertion that the input tensor is on host, is it missing here?

**Review by pilkicTT — COMMENTED** [review]
From what I see in the code, it doesn't look like the inputs to the program are allowed to be either on host or device (i.e. they must be always on host, if the binary expects it that way)? I couldn't find the place where the inputs to the program are specially handled if they're already on the device?

It would be good if we add tests for these kinds of basic API scenarios. Not necessarily with this PR...

File: runtime/lib/ttnn/runtime.cpp
Comment (pilkicTT) [inline runtime/lib/ttnn/runtime.cpp:511]: Is it just me, or would this be readable with some whitespace (new lines) in meaningful places, after scopes, group initialization of variables/consts, etc. 😄

File: runtime/lib/ttnn/include/CMakeLists.txt
Comment (jnie-TT) [reply runtime/lib/ttnn/include/CMakeLists.txt:1]: I'll add this a level up

File: runtime/lib/ttnn/runtime.cpp
Comment (jnie-TT) [reply runtime/lib/ttnn/runtime.cpp:367]: data shouldn't be referenced for ttnn tensors,there's no lifetime guarantee and we have it in the tensor. Also now that we support device tensors, they're going to be null pointer anyway for these tensors. I kept the member because metal runtime still uses it, but for consistency and safety I'm setting it to null across the board in ttnn runtime

File: runtime/lib/ttnn/runtime.cpp
Comment (jnie-TT) [reply runtime/lib/ttnn/runtime.cpp:511]: Yeah I was writing this at midnight everything was dark so wanted to keep it consistent to the environment therefore no whitespaces :)))). Jokes aside I'll add whitespaces to this and format it

**Comment by jnie-TT** [comment]
> From what I see in the code, it doesn't look like the inputs to the program are allowed to be either on host or device (i.e. they must be always on host, if the binary expects it that way)? I couldn't find the place where the inputs to the program are specially handled if they're already on the device?
> 
> It would be good if we add tests for these kinds of basic API scenarios. Not necessarily with this PR...

I'll add some runtime stitching tests that pre-moves tensors to device and run with tensors already on device, though right now the flatbuffer binaries have all input tensors hardcoded to host. We're currently looking at how we can canonicalize the input layout, would also be great to get your feedback there from a FE perspective.

File: runtime/lib/ttnn/runtime.cpp
Comment (pilkicTT) [reply runtime/lib/ttnn/runtime.cpp:367]: But isn't this borrowed data in ttnn? As long as the `Tensor` lives shouldn't we make sure that the `data` pointer lives?

Judging by the name `BorrowedStorage` I would assume that `ttnn` can never deallocate it by itself, since they are not owning it...

Comment (jnie-TT) [reply runtime/lib/ttnn/runtime.cpp:367]: The data pointer lifetime is managed by the user, we're currently borrowing the pointer with a non-tracking unsafe shared pointer that we create with unsafe_borrow_shared. Therefore having this pointer does not have any implications on lifetime, and for borrowed tensors we can access the data directly from the ttnn::tensor.

**Comment by jnie-TT** [comment]
@rpavlovicTT 
> Are you saying we'll be able to poll if tensor is ready/not?
Yes the APIs will block and poll for tensor readiness, and return only the info when the tensor is ready.

File: runtime/include/tt/runtime/runtime.h
Comment (nsmithtt) [inline runtime/include/tt/runtime/runtime.h:77]: Since the new submit API returns tensors maybe we should have convenience APIs:
```
void wait(Tensor tensor);
void wait(std::vector<Tensor> const& tensors);
```

File: runtime/lib/runtime.cpp
Comment (nsmithtt) [inline runtime/lib/runtime.cpp:200]: Next time for a big change, let's do cleanup/rename stuff like this as a separate PR, it makes reviewing the core change harder.

**Review by nsmithtt — APPROVED** [review]
Awesome, thank you for adding tests!  Just as a sanity check before we land, can we kick off a tt-forge-fe CI run rebased on this change?

Let's try to pre-empt any fallout.

**Review by kmabeeTT — APPROVED** [review]

**Pull request merged**
